Provide "Marshaling data" main article with explanation of how to call native code in C# (#24256)
dotnet/docs: docs/framework/interop/marshaling-data-with-platform-invoke.md

[!code-cpp[PInvokeLib#1](../../../samples/snippets/cpp/VS_Snippets_CLR/pinvokelib/cpp/pinvokelib.cpp#1)]

[!code-cpp[PInvokeLib#2](../../../samples/snippets/cpp/VS_Snippets_CLR/pinvokelib/cpp/pinvokelib.h#2)]

To call the library functions from managed code, first implement the managed prototypes for each function you want to invoke.
If the unmanaged code uses any any custom types, you must also declare those types in your managed code.  
Decorate the prototype with the <xref:System.Runtime.InteropServices.DllImportAttribute> attribute.

The following code shows an example prototype:

> ```cs

Reviewer: ~~~suggestion
```csharp
~~~

[!code-cpp[PInvokeLib#1](../../../samples/snippets/cpp/VS_Snippets_CLR/pinvokelib/cpp/pinvokelib.cpp#1)]

[!code-cpp[PInvokeLib#2](../../../samples/snippets/cpp/VS_Snippets_CLR/pinvokelib/cpp/pinvokelib.h#2)]

> To call this in a managed code (on the C# side) you should first implement the managed prototypes for all functions and methods which are to be invoked on the C# side.

Reviewer 2: ```suggestion
To call the library functions from managed code, first implement the managed prototypes for each function you want to invoke.
```

> If an unmanaged code uses a class or struct (or any custom type) this type has to be declared also in a managed code.  

Reviewer 2: ```suggestion
If the unmanaged code uses any custom types, you must also declare those types in your managed code.  
```

> Please see the <xref:System.Runtime.InteropServices.DllImportAttribute> which the prototype should be marked with.

Reviewer 2: ```suggestion
Decorate the prototype with the <xref:System.Runtime.InteropServices.DllImportAttribute> attribute.
```

> The example of such a prototype would look like:

Reviewer 2: ```suggestion
The following code shows an example prototype:
```

```cs

> // Declares a managed prototype for TestStructInStruct declared and defined in unmanaged library

Reviewer 2: ```suggestion
// Managed prototype for TestStructInStruct, which is declared and defined in an unmanaged library.
```

[DllImport("..\\LIB\\PinvokeLib.dll", CallingConvention = CallingConvention.Cdecl)]
internal static extern int TestStructInStruct(ref MyPerson2 person2);
```

> For more examples and comprehensive explanation please see other documents in this section related to marshaling.

Reviewer 2: ```suggestion
```

> 

Reviewer 2: ```suggestion
```

## See also

> 

Reviewer 2: ```suggestion

For more information and examples, see the following articles:

```
